WordPress/gutenberg
# Storybook: Set up local visual regression testing

## What?

Adds a setup to run visual regression tests against Storybook stories.

Some things to note:

- This is currently meant only as a development aid to use locally when doing complicated CSS refactors on specific components. Though, if it proves useful we may want to consider running it conditionally on CI when specific component file paths are modified.
- The tests run against a special build of Storybook, not the one we publish on https://wordpress.github.io/gutenberg. This is because generic stories are not particularly useful for vizreg testing. It's better that we write stories tailored for this purpose. Any story that matches `packages/components/src/**/e2e.stories.@(js|tsx|mdx)` will be included in this special Storybook build.
- The [`gotoStoryId()` util](https://github.com/WordPress/gutenberg/pull/43393/files#diff-2307e7879ecea1bcf61d376b2a1876c74617fece28dfbd0239a0f6af65c659d5R22) supports decorator options, so we can more easily run matrix tests involving things like LTR/RTL, global CSS, or margins.

## Why?

We've been wanting a more reliable way of testing visual concerns in components. Some unit tests currently use `.toHaveClass()` assertions or snapshot tests for this purpose, but they're not particularly reliable. The snapshot tests are especially not great because the rate of false positives are very high.

A vizreg setup like this will help us make CSS changes on complex components with more confidence. In particular, we will be starting cleanup work on the Button component (e.g. #44042), which will benefit from this kind of testing.

## Testing instructions

See the included [README](https://github.com/WordPress/gutenberg/pull/43393/files#diff-82ae6404a717e14e2ba66fb421ddb649b61cd7ca6e7b27cf6a3ee46a2f4f1d8b) for instructions.

## ✍️ Dev Note 

In order to have a more reliable way to find visual bugs that wouldn't otherwise be be discovered through other forms of testing, a first iteration of a Visual Regression testing framework has been implemented.

The framework relies on Playwright and Storybook, and it currently has a few limitations:
- it's mainly meant to test components from the `@wordpress/components` package;
- it only works locally (ie, the snapshots are not committed to the repository, and the tests can only be executed on a local installation of the Gutenberg repository).

Furthermore, this testing framework should be considered experimental — its setup may change in future releases of the Gutenberg plugin.

For more information on how to setup and run Visual Regression tests, read the [docs](https://github.com/WordPress/gutenberg/blob/trunk/test/storybook-playwright/README.md)
